### PR TITLE
Add items endpoints to GraphQL API and display items in database on frontend

### LINKED ISSUE #8: Show user data/authentication state and logout button on frontend

### PR DESCRIPTION
**Pull request by evan10s**
- **Database**
   - Added `categories` and `items` databases
   - No support yet for Google Cloud image storage
- **Backend**
   - Added item GraphQL endpoints (single item by ID or every item)
   - Restricted GraphiQL access to admins
- **Frontend**
    - Implemented Apollo Client for React
    - Frontend displays list of items stored in DB
- **General**   
   - Added in frontend authentication code and a few other missing items from the stale `graphql` branch
   - Cleaned up a whole bunch of style issues and removed some extraneous files

Closes #8

### CODE REVIEW

File: client/src/App.tsx [hunk 1]
```diff
@@ -17,35 +17,47 @@ import RequestManagementContainer from "./components/RequestManagementContainer"
...
+                            uuid
+                            name
+                            admin
+                          }
+                        }
+                    `
+            }),
+        });
+        const json = await userRequest.json();
+        const user = json.data.user;
```
Comment (joel99) [inline client/src/App.tsx]: non-null checks would be good

File: client/src/App.tsx [hunk 2]
```diff
@@ -17,35 +17,47 @@ import RequestManagementContainer from "./components/RequestManagementContainer"
...
     }
 
-    render() {
+
+    public render() {
         return (
             <div style={{
-                width: '100%',
+                width: "100%",
                 maxWidth: 960,
```
Comment (joel99) [inline client/src/App.tsx]: I would put `px` to be explicit

File: client/src/actions/actions.ts [hunk 3]
```diff
@@ -1,50 +1,50 @@
...
 }
 
 export interface LogoutUserAction {
-    type: string
+    type: string;
 }
 
 export const testAction: ActionCreator<Action> = () => ({
-    type: 'TEST_ACTION'
+    type: "TEST_ACTION"
```
Comment (joel99) [inline client/src/actions/actions.ts]: can we get rid of this

File: client/src/components/RequestManagementContainer.tsx [hunk 4]
```diff
@@ -91,25 +91,25 @@ class RequestManagementContainer extends Component<{}, RequestManagementContaine
         this.handleCross = this.handleCross.bind(this);
         this.handleReady = this.handleReady.bind(this);
         this.handlePrep = this.handlePrep.bind(this);
-        this.handleDone = this.handleDone.bind(this)
+        this.handleDone = this.handleDone.bind(this);
```
Comment (joel99) [inline client/src/components/RequestManagementContainer.tsx]: you can change class methods to arrow methods if you want to skip `bind`ing everything

File: client/src/components/inventory/HardwareList.tsx [hunk 5]
```diff
@@ -54,97 +57,138 @@ const sampleData = [
...
+                    noRequestsMessageText = "Sign in to request hardware.";
+                }
+
+                const noRequestsMessage = this.props.requestsEnabled && !this.props.user ? (<Message
+                    title="View-only inventory"
+                    warning icon>
+                    <Icon name="warning sign"/>
+                    {noRequestsMessageText}
+                </Message>) : "";
+                console.log(loading, error, data);
```
Comment (joel99) [inline client/src/components/inventory/HardwareList.tsx]: careful with logs on master, I don't mind deeply but don't forget

File: client/src/components/inventory/HardwareList.tsx [hunk 6]
```diff
@@ -54,97 +57,138 @@ const sampleData = [
...
+                    return <Message negative>
+                        <Message.Header>Error displaying hardware inventory</Message.Header>
+                        <p>Try refreshing the page. If that doesn't work, contact a member of the HackGT Team for
+                            assistance.</p>
+                    </Message>;
+                }
+
+                let noRequestsMessageText = "";
+
+                if (!this.props.requestsEnabled) {
```
Comment (joel99) [inline client/src/components/inventory/HardwareList.tsx]: can you move these checks to outside of the query and only perform query if necessary?

File: client/src/components/item/ItemEdit.tsx [hunk 7]
```diff
@@ -1,47 +1,47 @@
...
+            "loading": true,
+            "item": null,
             "item-return-required": true,
             "item-approval-required": true,
             "item-hidden": false,
-            itemOwnerChoices: [
-                {key: 'hackgt', text: 'HackGT', value: 'HackGT'},
-                {key: 'hive', text: 'The Hive', value: 'The Hive'},
-                {key: 'is', text: 'Invention Studio', value: 'Invention Studio'}]
+            "itemOwnerChoices": [
```
Comment (joel99) [inline client/src/components/item/ItemEdit.tsx]: can you extract these to a config file and import it instead
you can do this in a separate PR if you're fed up with this one

File: client/src/components/item/ItemEdit.tsx [hunk 8]
```diff
@@ -94,35 +94,35 @@ class ItemEdit extends Component<ItemEditProps, ItemEditState> {
...
             return {
-                key: item.replace(' ', '_').toLowerCase(),
+                key: item.replace(" ", "_").toLowerCase(),
                 value: item,
                 text: item
-            }
+            };
         });
 
         const itemOwnerChoices = [
```
Comment (joel99) [inline client/src/components/item/ItemEdit.tsx]: wait, this showed up earlier :thinking:

File: client/src/components/item/ItemEdit.tsx [hunk 9]
```diff
@@ -94,35 +94,35 @@ class ItemEdit extends Component<ItemEditProps, ItemEditState> {
...
 
         this.setState({
             [name]: value
         });
     }
 
-    render() {
-        const categories = ['Extension Cords', 'Laptops', 'Microcontrollers', 'Robotics'];
+    public render() {
+        const categories = ["Extension Cords", "Laptops", "Microcontrollers", "Robotics"];
```
Comment (joel99) [inline client/src/components/item/ItemEdit.tsx]: same here for config. probably do a single PR purging constants

File: client/src/index.tsx [hunk 10]
```diff
@@ -1,15 +1,50 @@
...
-import {Provider} from 'react-redux'
-import * as serviceWorker from './serviceWorker';
+import React from "react";
+import ReactDOM from "react-dom";
+import "./index.scss";
+import App from "./App";
+import {Provider} from "react-redux";
+import * as serviceWorker from "./serviceWorker";
 import {store} from "./store";
+import ApolloProvider from "react-apollo/ApolloProvider";
```
Comment (joel99) [inline client/src/index.tsx]: yikes, looks painful, gonna trust you did it rights

File: server/package.json [hunk 11]
```diff
@@ -21,6 +23,7 @@
     "@types/cookie-parser": "^1.4.1",
     "@types/cookie-signature": "^1.0.2",
```
Comment (joel99) [inline server/package.json]: these @types should be dev-dependencies

File: server/package.json [hunk 12]
```diff
@@ -8,7 +8,9 @@
     "start": "gulp",
     "serve": "node ./build/app.js",
     "build": "gulp build",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "generate-api": "node ./node_modules/@graphql-codegen/cli/dist/commonjs/cli.js --config graphql-gen.yml",
+    "prestart": "npm run generate-api"
   },
   "author": "Ryan Petschek",
```
Comment (joel99) [inline server/package.json]: you should add yourself as primary author :)

Comment (joel99) [reply server/package.json]: so people can hunt you down if there are bugs

File: server/api/api.graphql [hunk 13]
```diff
@@ -0,0 +1,8 @@
+type Query {
+	me: User
```
Comment (joel99) [inline server/api/api.graphql]: why are there two `api/api.graphql`s? Forget to delete one?

File: server/src/api/api.graphql [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	# Whether people who check out this item are expected to return it
+	returnRequired: Boolean!
+	# Whether admin approval is required before requests for this item will be processed
+	approvalRequired: Boolean!
+	# The person or organization that owns this item
+	owner: String!
+}
+
+input ItemInput {
+	item_name: String!
```
Comment (joel99) [inline server/src/api/api.graphql]: indent?

File: client/src/App.tsx (hunk 2 above)
Comment (evan10s) [reply client/src/App.tsx]: React assumes px if you don't put it, I remember getting a warning to remove px  (as it's unnecessary, or something) when I had it a while back.  I'll do some investigation though

File: client/src/components/RequestManagementContainer.tsx (hunk 4 above)
Comment (evan10s) [reply client/src/components/RequestManagementContainer.tsx]: Really?  WHOA

File: client/src/components/inventory/HardwareList.tsx (hunk 6 above)
Comment (evan10s) [reply client/src/components/inventory/HardwareList.tsx]: Didn't even think of it like that, great idea

File: client/src/components/item/ItemEdit.tsx (hunk 7 above)
Comment (evan10s) [reply client/src/components/item/ItemEdit.tsx]: It might need to be structured like categories TBH with a separate DB table tbh

File: client/src/components/item/ItemEdit.tsx (hunk 9 above)
Comment (evan10s) [reply client/src/components/item/ItemEdit.tsx]: Categories will be pulled from the server in the final version of this screen; this was just a placeholder

File: server/api/api.graphql (hunk 13 above)
Comment (evan10s) [reply server/api/api.graphql]: It's cruft.  I'll will make sure to delete

File: server/src/api/api.graphql (hunk 14 above)
Comment (evan10s) [reply server/src/api/api.graphql]: I don't see any issues here, can you be more specific?

File: client/src/components/inventory/HardwareList.tsx (hunk 6 above)
Comment (evan10s) [reply client/src/components/inventory/HardwareList.tsx]: Wait, nevermind.  Even if requests are disabled, we can still show the hardware list -- it has a special mode that hides the request buttons and only shows item details.  However, I did refactor this code slightly so it's nicer now

Comment (evan10s) [reply client/src/components/inventory/HardwareList.tsx]: image

File: server/api/api.graphql (hunk 13 above)
Comment (evan10s) [reply server/api/api.graphql]: I couldn't find this locally; otherwise I would have removed it

File: client/src/App.tsx (hunk 1 above)
Comment (evan10s) [reply client/src/App.tsx]: Done, although I want to refactor the fetch out of this component and into the action (or wherever the proper place is)

File: client/src/actions/actions.ts (hunk 3 above)
Comment (evan10s) [reply client/src/actions/actions.ts]: Done

File: client/src/components/RequestManagementContainer.tsx (hunk 4 above)
Comment (evan10s) [reply client/src/components/RequestManagementContainer.tsx]: Fixed this in a few files I came across, should check across the project to catch 'em all

File: client/src/components/inventory/HardwareList.tsx (hunk 5 above)
Comment (evan10s) [reply client/src/components/inventory/HardwareList.tsx]: Removed it

File: client/src/components/item/ItemEdit.tsx (hunk 7 above)
Comment (evan10s) [reply client/src/components/item/ItemEdit.tsx]: Probably separate PR material.  This is strongly tied to a separate discussion of whether we need to handle items that live in different locations and having multiple hardware desks (e.g., ours + one in The Hive), and item owner is kinda a surrogate for a proper Location field.

File: server/package.json (hunk 12 above)
Comment (evan10s) [reply server/package.json]: Done

File: client/src/components/item/ItemEdit.tsx (hunk 7 above)
Comment (joel99) [reply client/src/components/item/ItemEdit.tsx]: :ok:

File: server/api/api.graphql (hunk 13 above)
Comment (joel99) [reply server/api/api.graphql]: hmm, can you check the gitignore's and try force pushing to remove this? I see it on my end on a fresh pull of your branch

File: server/src/api/api.graphql (hunk 14 above)
Comment (joel99) [reply server/src/api/api.graphql]: it just appears differently, nbd

**Comment by evan10s** [comment]
@joel99 The superfluous `api.graphql` wasn't committed for some reason but also didn't show up in IntelliJ's commit window.  Shoud be gone now
